controllers/toolchaincluster/toolchaincluster_controller.go

```diff
@@ -0,0 +1,82 @@
+	"github.com/codeready-toolchain/toolchain-common/pkg/cluster"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	kubeclientset "k8s.io/client-go/kubernetes"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+// Reconciler reconciles a ToolchainCluster object
+type Reconciler struct {
+	Client     client.Client
+	Scheme     *runtime.Scheme
+	RequeAfter time.Duration
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *Reconciler) SetupWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&toolchainv1alpha1.ToolchainCluster{}).
+		Complete(r)
+}
+
+// Reconcile reads that state of the cluster for a ToolchainCluster object and makes changes based on the state read
+// and what is in the ToolchainCluster.Spec. It updates the status of the individual cluster
+// Note:
+// The Controller will requeue the Request to be processed again if the returned error is non-nil or
+// Result.Requeue is true, otherwise upon completion it will remove the work from the queue.
+func (r *Reconciler) Reconcile(ctx context.Context, request ctrl.Request) (ctrl.Result, error) {
+	reqLogger := log.FromContext(ctx)
+	reqLogger.Info("Reconciling ToolchainCluster")
+
+	// Fetch the ToolchainCluster instance
+	toolchainCluster := &toolchainv1alpha1.ToolchainCluster{}
+	err := r.Client.Get(ctx, request.NamespacedName, toolchainCluster)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			// Stop monitoring the toolchain cluster as it is deleted
+			return reconcile.Result{}, nil
+		}
+		// Error reading the object - requeue the request.
+		return reconcile.Result{}, err
+	}
+
+	cachedCluster, ok := cluster.GetCachedToolchainCluster(toolchainCluster.Name)
+	if !ok {
+		err := fmt.Errorf("cluster %s not found in cache", toolchainCluster.Name)
+		toolchainCluster.Status.Conditions = []toolchainv1alpha1.ToolchainClusterCondition{clusterOfflineCondition()}
+		if err := r.Client.Status().Update(ctx, toolchainCluster); err != nil {
+			reqLogger.Error(err, "failed to update the status of ToolchainCluster")
+		}
+		return reconcile.Result{}, err
+	}
+
+	clientSet, err := kubeclientset.NewForConfig(cachedCluster.RestConfig)
+	if err != nil {
+		reqLogger.Error(err, "cannot create ClientSet for the ToolchainCluster")
+		return reconcile.Result{}, err
+	}
+	healthChecker := &HealthChecker{
+		localClusterClient:     r.Client,
+		remoteClusterClient:    cachedCluster.Client,
+		remoteClusterClientset: clientSet,
+		logger:                 reqLogger,
+	}
+	//update the status of the individual cluster.
+	if err := healthChecker.updateIndividualClusterStatus(ctx, toolchainCluster); err != nil {
+		reqLogger.Error(err, "unable to update cluster status of ToolchainCluster")
+		return reconcile.Result{}, err
+	}
+
+	return reconcile.Result{RequeueAfter: r.RequeAfter}, nil
```

Comment:
I think I may have misunderstood the final decision 🙃 I was expecting something like the following:

```suggestion
       // execute healthcheck
       healthcheckResult, err := r.runHealthcheck(ctx, r.Client, cachedCluster.Client, clientSet, toolchainCluster, reqLogger)
       if err != nil {
		reqLogger.Error(err, "unable to run healthcheck for ToolchainCluster")
		return reconcile.Result{}, err
	}

	// update the status of the individual cluster.
	if err := r.updateStatus(ctx, healthcheckResult); err != nil {
		reqLogger.Error(err, "unable to update cluster status of ToolchainCluster")
		return reconcile.Result{}, err
	}

	return reconcile.Result{RequeueAfter: r.RequeAfter}, nil
```

Where `updateStatus` is a Reconciler function, and `runHealthcheck` is a Reconciler property that we can override in tests. So that we split responsibilities between the two components (healthchecker and reconciler) properly, and split tests accordingly. 

Is this change something we are planning to have? 😄

Reply:
Ok, it makes perfect sense. Thanks for clarifying

Reply:
Yeah, we agreed on moving the original code to the "controller" as it was and doing refactoring changes later in the following PR - to limit the number and the scope of the changes and to make reviews easier.